std/experimental/preallocated_array.d

```diff
@@ -0,0 +1,542 @@
+ */
+    ref Element opIndex(size_t i)
+    {
+        boundsCheck(i);
+        return store[i];
+    }
+
+/**
+Slicing operations execute an operation on an entire slice.
+
+Forwards everything to the underlying range.
+
+Precondition: $(D i < j && j < length)
+
+Complexity: $(BIGOH slice.length)
+ */
+    void opSliceAssign(Element value)
+    {
+        store[0 .. length] = value;
+    }
+
+    /// ditto
+    void opSliceAssign(Element value, size_t i, size_t j)
+    {
+        boundsCheck(j);
+        store[i .. j] = value;
+    }
+
+    /// ditto
+    void opSliceUnary(string op)()
+        if (op == "++" || op == "--")
+    {
+        mixin(op~"store[];");
+    }
+
+    /// ditto
+    void opSliceUnary(string op)(size_t i, size_t j)
+        if (op == "++" || op == "--")
+    {
+        mixin(op~"store[i .. j];");
+    }
+
+    /// ditto
+    void opSliceOpAssign(string op)(Elementvalue)
+    {
+        mixin("store[] "~op~"= value;");
+    }
+
+    /// ditto
+    void opSliceOpAssign(string op)(Elementvalue, size_t i, size_t j)
+    {
+        mixin("store[i .. j] "~op~"= value;");
+    }
+
+/**
+Forwards to $(LREF insertBack).
+ */
+    void opOpAssign(string op, Stuff)(Stuff stuff)
+        if (op == "~")
+    {
+        insertBack(stuff);
+    }
+
+/**
+Removes all contents from the container. 
+
+Postcondition: $(D empty)
+
+Complexity: $(BIGOH n)
+ */
+    void clear()
+    {
+        removeBack(length);
+    }
+
+/**
+Releases the underlying store. This PreallocatedArray
+instance is unuseable afterwarts.
+
+ */
+    Store release()
+    {
+        auto result = move(store);
+        return result[0 .. length];
+    }
+
+/**
+Picks one value in an unspecified position in the container, removes
+it from the container, and returns it. PreallocatedArray forwards removeAny to
+removeBack.
+
+Precondition: $(D !empty)
+
+Returns: The element removed.
+
+Complexity: $(BIGOH log(n)).
+ */
+    Element removeAny()
+    {
+        auto result = move(back);
+        removeBack();
+        return result;
+    }
+    /// ditto
+    alias stableRemoveAny = removeAny;
+/**
+Inserts $(D value) to the front or back of the container. $(D stuff)
+can be a value convertible to $(D Element) or a $(XREF2 range, isInputRange, range) of objects convertible
+to $(D Element). The stable version behaves the same, but guarantees that
+ranges iterating over the container are never invalidated.
+
+Params:
+    stuff = Either a $(XREF2 range, isInputRange, range) of elements to insert or an value convertible to $(D Element).
+
+Returns: The number of elements inserted
+
+Complexity: $(BIGOH m * log(n)), where $(D m) is the number of
+elements in $(D stuff)
+ */
+    size_t insertBack(Stuff)(Stuff stuff)
+        if (isImplicitlyConvertible!(Stuff, Element))
+    {
+        assert(_payload.refCountedStore.isInitialized);
+        if (length == capacity)
+            throw new RangeError("not enough room to insert stuff");
+        length = length + 1;
```

Comment:
what will happen if you have a stack allocated array and you try to increase the size.